### ISSUE TITLE
`missing-override-decorator` for python < 3.12 when `override` is imported via `typing_extensions`

### ISSUE THREAD

**Issue by michabirklbauer** [astral-sh/ty#3626]
### Summary

Hi, I am using `@override` in a project that targets python `< 3.12` (where `@typing.override` was introduced). I use `typing_extensions` for python versions `< 3.12` - `ty` does not seem to like that since I get `error[missing-override-decorator]` on any function that overrides another function even though every function has an `@override` decorator. For python `>= 3.12` and without `typing_extensions` it works as expected and does not complain.

Here is a minimal reproducible example / playground: https://play.ty.dev/eb521844-9ee5-4e38-8cbb-2b1c71b7f25f
```python
try:
    from typing import override  # ty: ignore[unresolved-import]
except ImportError:
    from typing_extensions import override


class Animal:
    name: str

    def __init__(self, name: str):
        self.name = name

    def get_name(self) -> str:
        return self.name


class Dog(Animal):
    name: str

    def __init__(self, name: str):
        self.name = name

    @override
    def get_name(self) -> str:
        return "Dog: " + self.name
```
with ty.toml:
```toml
[rules]
all = "error"

[environment]
python-platform = "all"
python-version = "3.10"
```

### Version

0.0.42

**Comment by lerebear** [comment]
Thank you for the report. This is a [known issue](https://github.com/astral-sh/ty/issues/1585) with this particular import pattern. In short, we support `typing_extensions.override`, but the dynamic nature of the import results in a type for `override` that is not definitive enough for us to recognize when evaluating the `missing-override-decorator` rule.

We don't yet have a timeline for that fix, but in the meantime perhaps one of these alternatives might work for you?

1. Perform a version check with `sys.version_info`:

```py
import sys

if sys.version_info >= (3, 12):
    from typing import override
else:
    from typing_extensions import override
```

2. Use `typing_extensions.override` for type checking specifically, but continue to use the dynamic import pattern at runtime:

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing_extensions import override
else:
    try:
        from typing import override
    except ImportError:
        from typing_extensions import override
```

**Comment by carljm** [comment]
Closing as duplicate of #1585, since I think this is an example of precisely what that issue already tracks. @lerebear lmk if you think there's value in keeping this open separately.

**Comment by michabirklbauer** [comment]
Thanks a lot for the clarification and for your help (and sorry for the duplicate!)

If I may ask a follow up question, I have used the try/except import pattern also for other types (or rather specifically for `Literal`) where this problem does not occur - but I think this maybe an edge case because `Literal` has been already introduced in `python 3.8` so it only affects code with python `< 3.8`. And it seems to come from `Literal` rather than the try/except import pattern.

Anyway, same pattern here does not produce an error, playground link https://play.ty.dev/35988aa7-dfd6-43ea-af66-1429e1cf9f17:
```python
try:
    from typing import Literal
except ImportError:
    from typing_extensions import Literal

a: Literal["a"] = "a"
```
ty.toml config:
```toml
[rules]
all = "error"

[environment]
python-platform = "all"
python-version = "3.7"
```
Python version is `==3.7` here which does not support `typing.Literal`:
```python
Python 3.7.17 (default, Sep 20 2023, 11:53:41)
[GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Literal
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'Literal' from 'typing' (/usr/local/lib/python3.7/typing.py)
>>>
```

However, this also is fine with ty - even though `typing.Literal` does not exist in python 3.7 and ty should raise an `unresolved import`:
https://play.ty.dev/233f42b0-094b-4d3a-a529-4c1389a7c998
```python
from typing import Literal

a: Literal["a"] = "a"
```
ty.toml config:
```toml
[rules]
all = "error"

[environment]
python-platform = "all"
python-version = "3.7"
```

Python 3.7 seems to be still support by ty, at least it is still listed here: https://docs.astral.sh/ty/reference/configuration/#python-version

**Comment by AlexWaygood** [comment]
We only partially support Python <3.10 at this point. We still allow you to pass `--python-version=3.7` on the command line or in a config file, and we will [still detect syntax errors](https://play.ty.dev/661899d1-5c71-47a7-a6fe-ba37382b3464) due to you using syntax features that were only added in 3.8, for example. However, our single source of truth for the standard library is the stdlib stubs that we vendor from [typeshed](https://github.com/python/typeshed). Typeshed has dropped support for Python <3.10, and so according to typeshed, `Literal` has always existed as a symbol in the `typing` module: https://github.com/astral-sh/ruff/blob/ec248a28b30f6f20091c3d9b6f31a0ba36f57186/crates/ty_vendored/vendor/typeshed/stdlib/typing.pyi#L492-L512.

**Comment by michabirklbauer** [comment]
Oh okay, I understand. Thank you very much for explaining!
